File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumRootToLeaf(TreeNode* root) {
        return solve(root, 0);
    }
private:
    int solve(TreeNode* root, int sum) {
        // Base case
        if(root == nullptr)
            return 0;
        //
        
        int rootValue = (sum << 1) | root-> val;
        if(root -> left == nullptr && root -> right == nullptr)
            return rootValue;
        
        return solve(root -> left, rootValue) + solve(root -> right, rootValue);
        
    }
};

File: 1022-sum-of-root-to-leaf-binary-numbers/1022-sum-of-root-to-leaf-binary-numbers.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def solve(self, root, currSum = 0):
        if root == None:
            return 0
        
        rootSum = (currSum << 1) + root.val
        
        if root.left == None and root.right == None:
            return rootSum
        
        return self.solve(root.left, rootSum) + self.solve(root.right, rootSum)
    
    def sumRootToLeaf(self, root: Optional[TreeNode]) -> int:
        return self.solve(root)
    